generator/translator_geocoder_base.cpp

```diff
@@ -0,0 +1,154 @@
+#include "generator/translator_geocoder_base.hpp"
+
+#include "generator/collector_interface.hpp"
+#include "generator/emitter_interface.hpp"
+#include "generator/feature_builder.hpp"
+#include "generator/generate_info.hpp"
+#include "generator/holes.hpp"
+#include "generator/intermediate_data.hpp"
+#include "generator/osm2type.hpp"
+#include "generator/osm_element.hpp"
+#include "generator/regions/collector_region_info.hpp"
+
+#include "indexer/classificator.hpp"
+
+#include "geometry/point2d.hpp"
+
+#include "base/assert.hpp"
+#include "base/geo_object_id.hpp"
+
+#include <set>
+#include <string>
+#include <vector>
+
+namespace generator
+{
+TranslatorGeocoderBase::TranslatorGeocoderBase(std::shared_ptr<EmitterInterface> emitter,
+                                               cache::IntermediateDataReader & holder)
+  : m_emitter(emitter), m_holder(holder) {}
+
+void TranslatorGeocoderBase::EmitElement(OsmElement * p)
+{
+  CHECK(p, ("Tried to emit a null OsmElement"));
+
+  FeatureParams params;
+  // We are forced to copy, because ParseParams() is a bad function,
+  // because it removes tags from OsmElement.
```

Comment:
Зачем здесь эмоциональная окраска? Если функция плохая, надо её или править, или писать todo непосредственно к ней.
```
// We are forced to copy because ParseParams() removes tags from OsmElement.
```

Reply:
++